### DISCUSSION TITLE
Best way to generate A3 from A4

### DISCUSSION THREAD

**Discussion by KillTheMule** [typst/typst#2121] [General]
I'm creating documents of the format A4 for visual consumption, but I need them on A3 for printing. Note that this doesn't mean upscaling them, but putting 4 A4 documents in the right order on both sides of a A3 document (2 each side). I'm generating the documents programatically from templates, and I really need to make sure that the A3 is exactly occupied by the A4 documents. Moreover, I'd prefer if it was easy to keep those documents in sync...

Up to now, I've been using latex for this, and producing the A4 was really just including the A4 pdf's in the right order (using the package `pdfpages`). I want to switch to typst, which can't include pdf's right now. So, what's my best option here? I do of course know the `columns` function, but I do foresee problems with the page headers/footers and page numbers at least...

Thanks for any pointers! Let me know if anything was unclear here :)

**Comment by laurmaedje** [comment]
There isn't a good built-in way for this at the moment. Your best bet is probably staying completely in A4 in Typst, and then converting the output to A3 with another tool.

**Reply by KillTheMule** [reply]
Hurr, that's bad, seeing that my main motivation for switching right now is to get away from needing a latex installation. I'll shop around for a rust lib for pdf manipulation \*fingers crossed\* Thanks!

**Reply by KillTheMule** [reply]
Wait a second, I've been looking around and been considering using a pdf output library that can use png which typst could ouput... but then again, typst has `#figure`. Could I not render my a4 documents into a png with typst and then have typst include that in the a3 document? Or am I missing something?

**Reply by laurmaedje** [reply]
Going through PNG would probably work if that works for you. Maybe even SVG then (which is availble on main and also in the next release that is coming in the next days).

**Reply by KillTheMule** [reply]
Gonna give that a shot, thanks!

**Comment by alerque** [comment]
This is probably *not* something `typst` should do directly, unless you *want* it to become as bloated as TeX ;-) In all seriousness this is a fundamentally different operation than typesetting and *should* be part of specialty tooling as a post-processor.

You can great N-up layouts like this for printing with a wide variety of tools that don't involve a full TeXLaTeX installation (or any at all). Try `pdftk`, `qpdf`, `gs`, `pypdf`, or any one of the small army or alternatives both offline as CLI tools, as visual interfaces to CLI tools, or online. Also most printer drivers support N-up printing anyway.

**Reply by KillTheMule** [reply]
Sure, there are tools, but my goal is to not have my users install something other than my program, and I need the whole thing to work cross-plattform (linux, windows). I'd understand if this is out of scope for typst though :)

**Reply by alerque** [reply]
Then IMO "your tool" should include typst as a library and also include one of the PDF manipulation libraries that handles creating new layouts and add your own routine to capture the output and setup a print layup.

**Reply by KillTheMule** [reply]
> ... include one of the PDF manipulation libraries that handles creating new layouts 

Yeah, pretty much what I said above. But, any suggestion about a lib? Needs to be useable from rust and be mit/apache licensed, and I have not yet found a candidate. I pondered including a qpdf binary, writing that out and running that, but seems like qpdf can't handle this. Poppler could, I think, but I can't use a gpl lib.

**Reply by alerque** [reply]
So not only do you want other people to do your research for you, you're hoping other people will have written code for you already AND licensed in a way that you don't have to share back and/or can charge your downstream customers for the convenience of a a closed source final product. Sign me out.

**Reply by KillTheMule** [reply]
Woa, not sure where that came from, the conversation seemed somewhat polite, and suddenly asking a question is too much? I mean, you're not required to answer anything of course, but you've chosen to join in yourself, and brought up the idea of not trying to use typst here, so it didn't feel too weird to ask if you had more input on the matter...

**Reply by laurmaedje** [reply]
@alerque You are free to not reply to a question, but if do, please stay friendly and constructive.

**Comment by KillTheMule** [comment]
So, I'd like to reopen this to report my findings, maybe someone finds them usefull:

1. If investigated using [pdfium-rs](https://github.com/asafigan/pdfium_rs), which mostly fits the bill. However, I could not make pdfium (the c++ library this wraps) link statically on windows, so I had to give up. A drawback of this approach is that this cannot be used with multiple threads properly.
2. I've realized that my use case is not quite "recombine A4 pdfs into A3 pdfs", but really "recombine typst A4 documents into typst A3 documents", which is actually solvable quite nicely with the following code (where `doc` is the A4 document produced by typst):
```rust
let page1 = &doc.pages[0];
let page2 = &doc.pages[1];

let frame1 = &page1.frame;
let frame2 = &page2.frame;

let xdim = frame1.size().x;

let newsize =  typst::layout::Size { x: 2.0 * xdim, y: frame1.size().y };
let mut frame = typst::layout::Frame::new(newsize, frame1.kind());
frame.set_baseline(frame1.baseline());


for item1 in frame1.items() {
  let it = item1.clone();
  frame.push(it.0, it.1);
}

for item2 in frame2.items() {
  let item = item2.1.clone();
  let point = item2.0;
  let newpoint = typst::layout::Point { x: point.x + xdim, y: point.y };
  frame.push(newpoint, item);
}

let page = typst::layout::Page {
  frame,
  numbering: None,
  number: 0
};

let v = vec![page];
doc.introspector.rebuild(&v);
doc.pages = v;
```

This merges two A4 pages into one A3 page, but the rest is just a bit of bookkeeping I'll have to write (I need to go 4 A4 -> 2 A3 and similar), and then after all merge it into one large A3 document. Note that some choices here are specific to my usecase (e.e. reusing the old `doc`, which keeps things like the title, author etc), so this doesn't quite work as a general document-merge.

As far as I can tell though, this does the job very properly, but if anyone spots something I might have missed, I'd be grateful for any hint :) Otherwise, I hope this helps someone somewhere somehow!

(e) I'll investigate if https://docs.rs/typst/latest/typst/layout/struct.Frame.html#method.push_frame is more efficient (since I can only clone the individual items, this method might just transfer ownership), but that's just a performance optimization after al...

**Comment by KillTheMule** [comment]
So here's my final version (error handling removed since it's pretty specific). It's A LOT faster than going through png :)

```rust
pub fn merge_frames(frame1: Option<Frame>, frame2: Frame) -> Frame {
  let xdim = frame2.size().x;
  let ydim = frame2.size().y;

  let newsize = Size::new(2.0 * xdim, ydim);
  let mut newframe = Frame::new(newsize, frame2.kind());
  newframe.set_baseline(frame2.baseline());

  if let Some(frame1) = frame1 {
    newframe.push_frame(Point::zero(), frame1);
  }

  newframe.push_frame(Point::new(xdim, Abs::zero()), frame2);

  newframe
}

// Combines 2 pages of the document into one page of double size. Returns an
// error if the document does not have 3 or 4 pages.
pub fn create_a3(mut doc: Document) -> Document {
  let mut pages = doc.pages;
  let len = pages.len();

  let page1 = match len {
    n @ 0..=2 => panic!("Can't merge"),
    3 => None,
    4 => pages.pop(),
    n @ 5.. => panic!("Can't merge"),
  };

  let [page2, page3, page4] = pages.try_into().unwrap();

  let frame1 = merge_frames(page1.map(|p| p.frame), page2.frame);
  let frame2 = merge_frames(Some(page3.frame), page4.frame);

  let page1 = Page {
    frame: frame1,
    numbering: None,
    number: 0,
  };
  let page2 = Page {
    frame: frame2,
    numbering: None,
    number: 0,
  };

  let v = vec![page1, page2];
  doc.introspector.rebuild(&v);
  doc.pages = v;

  Ok(doc)
}
```
